src/main.rs

```diff
@@ -1,3 +1,176 @@
+        }
+
+        if console_output.contains("=>") || timeout_counter >= 100 {
+            break;
+        }
+    }
+    send(serial_port, b"\x03").expect("Failed to write to serial port"); // clear prompt
+    console_output
+}
+
+fn run_u_boot_cmd(serial_port: &mut Box<dyn SerialPort>, cmd: &str) -> String {
+    send(serial_port, format!("{cmd}\n").as_bytes()).expect("Failed to write to serial port");
+
+    let mut console_output = String::new();
+    let mut timeout_counter = 0;
+
+    loop {
+        if let Some(s) = receive(serial_port) {
+            console_output += s.as_str();
+            timeout_counter = 0;
+        } else {
+            timeout_counter += 1;
+        }
+
+        if timeout_counter >= 10 {
+            break;
+        }
+    }
+    console_output
+}
+
+fn report_issue(issue: &str, instructions: &str) {
+    println!("! {issue}");
+    println!("-> {instructions}");
+}
+
+fn run_u_boot_check(
+    serial_port: &mut Box<dyn SerialPort>,
+    cmd: &str,
+    pattern: &str,
+    issue: &str,
+    instructions: &str,
+) -> bool {
+    let console_output = run_u_boot_cmd(serial_port, cmd);
+
+    if !console_output.contains(pattern) {
+        report_issue(issue, instructions);
+        return false;
+    }
+    true
+}
+
+fn analyze(serial_port: &mut Box<dyn SerialPort>) {
+    let early_check_info = vec![
+        ("U-Boot SPL", "No U-Boot detected"),
+        ("DRAM:  128 MiB", "Wrong RAM size detected"),
+        (
+            "Net:   eth0: eth@10110000",
+            "Ethernet could not be initialized",
+        ),
+        ("=>", "Could not enter U-Boot shell"),
+    ];
+    let console_output = enter_u_boot(serial_port);
+
+    for (pattern, issue) in early_check_info {
+        if !console_output.contains(pattern) {
+            report_issue(issue, INSTRUCTIONS_LM);
+            return;
+        }
+    }
+
+    let u_boot_check_info = vec![
+        (
+            "mtd list",
+            "spi-nand0",
+            "NAND flash not detected",
+            INSTRUCTIONS_LM,
+        ),
+        (
+            "gpio input PA11",
+            "gpio: pin PA11 (gpio 11) value is 1",
+            "Button stuck",
+            INSTRUCTIONS_BUTTON,
+        ),
+    ];
+
+    for (cmd, pattern, issue, instructions) in u_boot_check_info {
+        if !run_u_boot_check(serial_port, cmd, pattern, issue, instructions) {
+            return;
+        }
+    }
+
+    println!("! No issues found");
+}
+
 fn main() {
-    println!("Hello, world!");
+    let mut serial_port_name = String::new();
+
+    if let Ok(ports) = available_ports() {
+        if ports.len() == 1 {
+            serial_port_name = ports[0].port_name.clone();
+        } else {
+            let choices = ports.into_iter().map(|p| p.port_name).collect();
+            serial_port_name = inquire::Select::new("Select serial port", choices)
+                .prompt()
+                .expect("Invalid selection");
+        }
+    }
```

Comment:
I try to avoid `mut` as much as possible

```suggestion
    let serial_port_name = if let Ok(ports) = available_ports() {
        if ports.len() == 1 {
            ports[0].port_name.clone()
        } else {
            let choices = ports.into_iter().map(|p| p.port_name).collect();
            inquire::Select::new("Select serial port", choices)
                .prompt()
                .expect("Invalid selection")
        }
    } else {
        // or return error instead?
        String::new()
    };
```